pallets/worker/src/lib.rs

```diff
@@ -372,128 +374,41 @@ decl_module! {
-                // right.
-                // we will will generate the next epoch dataset and store it in the other end of
-                // the dataset, also we will lock here to avoid double generation.
-
-                // this the same lock used before to generate the first dataset, but we are 100%
-                // sure that the we have a dataset, so there will be no locking there.
-                let mut lock = StorageLock::<Time>::new(constants::storage_keys::DAG_DATASET_LOCK);
-                if let Ok(_guard) = lock.try_lock() {
-                    // if we are using the left side, we should store the next DAG Dataset in the
-                    // right, and the other way around.
-                    let key = if use_left {
-                        StorageValueRef::persistent(constants::storage_keys::RIGHT_DAG_DATASET)
-                    } else {
-                        StorageValueRef::persistent(constants::storage_keys::LEFT_DAG_DATASET)
-                    };
-                    let next_epoch = stored_epoch + 1;
-                    // this will generate and store the dataset in the key.
-                    generate_dataset(next_epoch, cache.as_slice(), &key);
-                };
-            }
             let latest_block_number = Self::last_block_number();
-            let call = if U256::from(header.number) > latest_block_number {
+            let call = if header.number > latest_block_number {
                 if Self::initialized() {
-                    // TODO(drewstone): we should convert this proofs to DoubleNodeWithMerkleProof
-                    // but how?
-                    let _tree = ethash::calc_dataset_merkle_proofs(
-                        stored_epoch as usize,
-                        dataset.as_slice(),
-                    );
-                    let proof = [DoubleNodeWithMerkleProof::new()].to_vec();
-                    debug::native::info!("Adding header #: {:?}!", header.number);
-                    Some(Call::add_block_header(rlp_header, proof))
+                    if let Some(proofs) = generate_proofs(&rlp_header) {
+                        debug::native::info!("Adding header #: {:?}!", header.number);
+                        Some(Call::add_block_header(rlp_header, proofs))
+                    } else {
+                        // it sounds that we can't generate the proofs for that header now.
+                        // push it back to the front of the queue.
+                        header_queue.push_front(header);
+                        None
+                    }
                 } else {
                     debug::native::info!("Initializing with header #: {:?}!", header.number);
-                    let dags_start_epoch: u64 = header.number / 30000;
+                    let dags_start_epoch = header.number.as_u64() / 30000;
                     Some(Call::init(
                         dags_start_epoch,
                         vec![],
```

Comment:
Maybe add roots here too?

Reply:
yeah, I guess we could add the roots as hardcoded static array in `constatnts` module.